File: javascript/triangle/triangle.js
'use strict';

function Triangle() {
  const sides = Array.prototype.slice.call(arguments, 0, 3);

  //avoiding mutating sides
  const sortedSides = Array.from(sides).sort();
  const types = { 1:'equilateral', 2:'isosceles', 3:'scalene' };

  const isAnInvalidSideSize = (a) => a <= 0;
  const violatesTriangleInequality = (sides) => sides[0] + sides[1] < sides[2];

  const kind = () => {
    if (sides.some(isAnInvalidSideSize)) {
      throw new RangeError(`Triangle (${String(sides)}) has invalid side size`);
    }

    if (violatesTriangleInequality(sortedSides)) {
      throw Error(`Triangle (${String(sides)}) violates triangle inequality`);
    }

    return types[new Set(sides).size];
  };

  return Object.freeze({ kind });
}

module.exports = Triangle;

File: python/triangle/triangle.py
TRIANGLES = {1: 'equilateral', 2: 'isosceles', 3: 'scalene'}


def violatesTriangleInequality(sides):
    (a, b, c) = tuple(sides)
    return a + b <= c


class Triangle:
    def __init__(self, *args):
        sides = sorted(args)
        if (violatesTriangleInequality(sides)):
            raise TriangleError(
                "Triangle (${0}) violates triangle inequality".format(sides))

        if (any([side <= 0 for side in sides])):
            raise TriangleError("Triangle (${0}) has invalid side size".format(
                sides))
        self.sides = sides

    def kind(self):
        return TRIANGLES.get(len(set(self.sides)))


class TriangleError(Exception):
    pass
